### PR TITLE
feat(deploy): make diamondCut calldata recomputable from main (EXSC-855)

### PR DESCRIPTION
**Pull request by 0xDEnYO**
# Which Linear task belongs to this PR?

Fixes [EXSC-855](https://linear.app/lifi-linear/issue/EXSC-855/w11-recompute-diamondcut-calldata-from-main-harness)

Sub-issue of [EXSC-686 · Multisig Signing Process 2.0](https://linear.app/lifi-linear/issue/EXSC-686/multisig-signing-process-20). This is work package **W1.1 — the harness only**. The comparison engine in `confirm-safe-tx.ts` is W1.2 and ships separately.

# Why did I implement it this way?

At signing time nothing verifies that a Safe proposal's `diamondCut` calldata is what `main`'s own scripts would produce from config on `main`. Leg 1 of the fix is "recompute from main and require byte-for-byte equality"; this PR makes that recomputation possible.

`UpdateScriptBase` already had most of the harness (`NO_BROADCAST=true` returns the full `cutData` and skips `vm.startBroadcast`). What it did not have was a way to stop trusting three inputs that whoever prepared the proposal chose:

| Input | Was | Now |
| --- | --- | --- |
| Facet address | `deployments/<network>.<suffix>json` | `FACET_ADDRESS_OVERRIDE` — the caller passes the address the proposal claims. Proving that address holds the right bytecode is a separate concern (S20 / EXSC-706), deliberately not solved here. |
| Diamond address | same file | `EXPECTED_DIAMOND_ADDRESS` pins it, and every run now asserts `block.chainid` against `config/networks.json`. |
| Selectors | whatever sat in local `out/` | `SELECTOR_ARTIFACTS_DIR`, passed explicitly to `contract-selectors.sh` instead of the script hardcoding `./out`. |

### The diamond cross-check is against chainId, not a diamond address

The ticket asked for a cross-check against `config/networks.json`. **That file carries no diamond address** — I checked every key across all entries, and the per-network `LiFiDiamond` addresses in `deployments/` are not uniform either, so there is nothing there to compare to. What `networks.json` *can* authorise is the chain: `_checkDiamondAddress()` reverts `NetworkChainIdMismatch` when `block.chainid` differs from `.<network>.chainId`. That is the failure this check actually needs to catch — a recomputation aimed at the wrong RPC silently produces a completely different, wrong cut. Networks absent from `networks.json` skip the assertion rather than blocking.

Pinning the diamond itself is therefore an explicit caller input (`EXPECTED_DIAMOND_ADDRESS`), which is honest about where the trust comes from.

### The cut is not a pure function of main

`buildDiamondCut()` queries the live diamond through the loupe to decide Add/Replace/Remove. If another cut lands between propose and verify, an honest proposal legitimately stops matching. So `CUT_VERIFICATION_MODE=true`:

- requires `DIAMOND_STATE_BLOCK`, reverting `DiamondStateNotPinned` when it is absent;
- reverts `DiamondStateBlockMismatch(expected, actual)` when the fork is not actually at that height (i.e. the caller forgot `--fork-block-number`, or the diamond moved on);
- reverts `DiamondHasNoCode` when the pinned diamond has no bytecode;
- forces `noBroadcast`, regardless of what the caller passed.

These are named, distinct conditions: "the comparison could not be made reproducibly" is not the same answer as "the calldata differs", and W1.2 has to be able to tell them apart.

### Default behaviour

Unchanged, and pinned by tests rather than asserted:

- `test_ReplaceCutMatchesGoldenCalldata` / `test_AddCutMatchesGoldenCalldata` assert the exact `cutData` bytes for both cut shapes against a diamond built in-test.
- `test_ExplicitDefaultOptionsProduceIdenticalCalldata` asserts that setting the new options to their defaults produces identical bytes.
- I also diffed `contract-selectors.sh` against `origin/main`'s copy over four facets, with and without excludes — byte-identical output when the new third argument is omitted (it defaults to `./out`, or `./out/zksync` for the zkSync copy).

The only intentional behaviour change on the default path is the new `chainid` assertion, which can only fire when the run is pointed at the wrong chain.

### Fail-open gaps found by the review gate and closed

Three defects in the first commit, all of the same shape — a safety check that degrades to "off" without saying so. Fixed in 9b26200be:

- **`vm.envOr` swallows a set-but-unparseable value** and returns the default. A typo in `CUT_VERIFICATION_MODE` silently disabled the block pin and the forced no-broadcast; a typo in `FACET_ADDRESS_OVERRIDE` silently produced plausible calldata built from the deployments file — the exact source the override exists to displace. The five new variables now read strictly (`vm.envExists` plus the typed getter), so a malformed value reverts. `NO_BROADCAST` deliberately keeps its existing `envOr` semantics so no current flow changes.
- **`contract-selectors.sh` only guarded a missing file.** A malformed or truncated artifact, or an excludes list covering every selector, still exited 0 with an empty `bytes4[]` — which encodes as a valid *no-op diamond cut*, not an error. That matters more now that `SELECTOR_ARTIFACTS_DIR` points the script at foreign build trees. All four paths now exit 1; I verified each against real artifacts, and confirmed no facet's real `getExcludes()` covers its whole selector set.
- **`UpdateCoreFacets.s.sol` and `UpdateDiamondLoupeFacet.s.sol` resolve facet addresses themselves**, not through `update()`, so the overrides never reached them while all four new reverts did — a verification run there looked fully armed while sourcing every address from local state. Both now revert `VerificationModeNotSupported`.

The gate also showed the first commit's tests covered none of the env var *names* — every harness overrode the options in Solidity, so a mistyped literal would have passed the suite. The suite now drives all five from the real env, and adds the missing `DiamondHasNoCode` case and a golden for `buildDiamondCut`'s Remove branch.

### Two things a reviewer should push back on if they disagree

- **`_readCutOptions()` is `virtual`.** forge shares process env across test cases it runs in parallel, so driving each case by mutating env vars is flaky by construction (I hit exactly that — 8 of 11 tests failed multi-threaded, all 11 passed at `-j 1`). Making the env read a single overridable seam is what lets the suite cover every combination deterministically. It also gives W1.2 one place to look for the full option set.
- **`ScriptBase` still demands `PRIVATE_KEY`** even in a no-broadcast verification run, because it derives `deployerAddress` in its constructor. I did not relax that — loosening it to an `envOr` default would let a real broadcast flow run with a zero key. The doc says to pass a throwaway key instead.

zkSync copies of `UpdateScriptBase.sol`, `contract-selectors.sh` and `UpdateCoreFacets` carry the identical change, per `.agents/rules/107-solidity-scripts.md`. The two `UpdateScriptBase.sol` copies differ only in the artifacts-dir default, the `cmd[0]` script path, and a pre-existing `virtual` on `getSelectors`.

### Open, escalated to you

Two items from the review gate are **not** fixed and need your decision — see [the gate comment](https://github.com/lifinance/contracts/pull/2275#issuecomment-5420856060): whether verification mode should *require* the overrides rather than merely accept them, and a pre-existing bug where `UpdateDiamondLoupeFacet.s.sol` broadcasts before checking `noBroadcast`.

### Tests / lint run

- `forge test --match-path test/solidity/script/UpdateScriptBase.t.sol` — 14 passed, run repeatedly multi-threaded to confirm the env-race is gone.
- `forge test --no-match-path "test/solidity/Facets/**"` — 121 passed, 24 failed. **Zero** of the failures are from this change: all 24 are `vm.envString: environment variable "ETH_NODE_URI_*" not found` in `setUp()`, because this worktree has no RPC env, and the same 24 fail on an unmodified tree. Fork-backed facet suites were not run locally for the same reason — CI covers them.
- `bunx solhint` on all three changed/added Solidity files — 0 errors.
- `bash -n` on both shell scripts, plus a live dry-run of the new missing-artifact guard.
- `bunx markdownlint-cli2` on both docs — 0 errors.

# Checklist before requesting a review

- [x] I have performed a self-review of my code
- [x] This pull request is as small as possible and only tackles one problem
- [x] I have added tests that cover the functionality / test the bug
- [ ] For new facets: I have checked all points from this list: https://www.notion.so/lifi/New-Facet-Contract-Checklist-157f0ff14ac78095a2b8f999d655622e
- [x] I have updated any required documentation

# Checklist for reviewer (DO NOT DEPLOY and contracts BEFORE CHECKING THIS!!!)

- [ ] I have checked that any arbitrary calls to external contracts are validated and or restricted
- [ ] I have checked that any privileged calls (i.e. storage modifications) are validated and or restricted
- [ ] I have ensured that any new contracts have had AT A MINIMUM 1 preliminary audit conducted on <date> by <company/auditor>

### CODE REVIEW

**Comment by 0xDEnYO** [comment]
## Review gate — escalated items (need a decision before this goes ready)

Auto-fixable findings were fixed in 9b26200be. These two were not, and need your call.

### 1. Verification mode does not require the overrides its own design says are the trustworthy sources

`CUT_VERIFICATION_MODE=true` currently requires only `DIAMOND_STATE_BLOCK`. A run with the mode on and no `FACET_ADDRESS_OVERRIDE`, no `EXPECTED_DIAMOND_ADDRESS` and no `SELECTOR_ARTIFACTS_DIR` succeeds and recomputes entirely from the local `deployments/<network>.json` and local `./out` — precisely the proposer-written input the mode exists to displace. It will "match" a proposal whenever the local checkout already holds the proposed address, which is the normal case for whoever prepared it.

I did not change this because it fixes W1.2's calling contract, and the three overrides are not equally clear-cut:

- `FACET_ADDRESS_OVERRIDE` and `EXPECTED_DIAMOND_ADDRESS` — I'd require both. A verification run without them is the false-assurance case with no legitimate use I can see.
- `SELECTOR_ARTIFACTS_DIR` — arguable. A verifier who controls the checkout and built `main` into the default `./out` is legitimately using main's artifacts, and forcing an explicit path buys nothing there.

For now `docs/DiamondCutRecomputation.md` says plainly that the mode is not self-sufficient, so nothing over-claims while this is open. **Decision needed: require all three, require the first two, or leave it to the caller.**

### 2. Pre-existing: `UpdateDiamondLoupeFacet.s.sol` broadcasts before it checks `noBroadcast`

`script/deploy/facets/UpdateDiamondLoupeFacet.s.sol` runs `vm.startBroadcast` → `cutter.diamondCut` → `vm.stopBroadcast` unconditionally when the loupe is absent, and only tests `noBroadcast` afterwards. So `NO_BROADCAST=true` does not actually suppress that cut. This predates this PR and is out of its scope; this PR only stops verification mode from reaching that script (`VerificationModeNotSupported`). **Worth its own ticket — flagging rather than fixing here.**

### Lower confidence, human judgment

- `_checkDiamondAddress` reads `config/networks.json` at a fixed path while `script/helperFunctions.sh` honours `NETWORKS_JSON_FILE_PATH`. I kept the fixed path deliberately (a trust anchor an env var can redirect is not one) and said so in a comment, but the asymmetry is worth a second opinion.
- `contract-selectors.sh`'s `${3:-./out}` default is now unreachable from the only in-repo callers, since both always pass all four arguments. Kept so the script stays usable standalone.

File: docs/DiamondCutRecomputation.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# Recomputing diamondCut calldata
+
+`UpdateScriptBase` can rebuild the `diamondCut` calldata for a facet without broadcasting, so a
+signer can compare a pending Safe proposal against the calldata this repository's own scripts
+produce from the config on `main`. Nothing here changes how a normal deploy or proposal run
+behaves: with every variable below unset, the script runs exactly as before.
```
[the remark this answers is not in the document]
Comment (0xDEnYO) [reply docs/DiamondCutRecomputation.md]: Valid — fixed in 7a6a5d031. The opening now states that `_checkDiamondAddress()` runs on every path and that a configured network whose RPC chain ID mismatches reverts `NetworkChainIdMismatch`, rather than claiming the unset configuration is unchanged.

**Comment by 0xDEnYO** [comment]
## Escalated items — resolved, PR flipped to ready

Following up on the review-gate escalation above.

**1. Verification mode did not require its own trust anchors** — fixed in `41faf5cbf`. Took the middle option: `FACET_ADDRESS_OVERRIDE` and `EXPECTED_DIAMOND_ADDRESS` are now both required in verification mode (`FacetAddressOverrideRequired` / `ExpectedDiamondAddressRequired`), while `SELECTOR_ARTIFACTS_DIR` stays optional — a verifier who built `main` into the default `out/` with the network's compiler group is using a trustworthy tree. `docs/DiamondCutRecomputation.md` gained a per-group build table so "the right `out/`" is unambiguous.

Re-gated the two commits pushed after the original gate (`7a6a5d031`, `41faf5cbf`), since gate clearance covers only the commits it reviewed:

- Ordering verified: `_readCutOptions()` populates `cutOptions` in the constructor (line 65) before `_checkDiamondStateIsPinned()` runs (line 89), so the guards cannot false-positive on a correctly-invoked run.
- Firing verified: the unset-env path is `vm.envExists(...) ? vm.envAddress(...) : address(0)`, so an omitted override genuinely reaches the guard as zero. Worth noting the same-PR tests override `_readCutOptions` (deliberately — forge shares process env across parallel tests), so they are not on their own evidence that the real env path can produce zero; the ternary above is.
- Paired representation: the zkSync `UpdateScriptBase.sol` carries the identical change.

**2. `UpdateDiamondLoupeFacet.s.sol` broadcasts before it checks `noBroadcast`** — filed as [EXSC-891](https://linear.app/lifi-linear/issue/EXSC-891/updatediamondloupefacetssol-broadcasts-the-loupe-cut-before-it-checks). Pre-existing and out of scope here; this PR only stops verification mode from reaching that script, so the plain `NO_BROADCAST=true` path is still affected and needs its own fix.

The two lower-confidence items from the original comment (`_checkDiamondAddress` reading `config/networks.json` at a fixed path, and `contract-selectors.sh`'s now-unreachable `${3:-./out}` default) are unchanged and still worth a reviewer's second opinion.

**Comment by 0xDEnYO** [comment]
@coderabbitai review

**Review by gvladika — APPROVED** [review]
